bindings/src/main/java/io/pravega/storage/extendeds3/ExtendedS3Storage.java

```diff
@@ -386,23 +388,57 @@ private void setPermission(SegmentHandle handle, Permission permission) {
      * completeMultiPartUpload call. Specifically, to concatenate, we are copying the target segment T and the
      * source segment S to T, so essentially we are doing T <- T + S.
      */
-    private Void doConcat(SegmentHandle targetHandle, long offset, String sourceSegment) throws StreamSegmentNotExistsException {
+    private Void doConcat(SegmentHandle targetHandle, long offset, String sourceSegment) throws Exception {
         Preconditions.checkArgument(!targetHandle.isReadOnly(), "target handle must not be read-only.");
         long traceId = LoggerHelpers.traceEnter(log, "concat", targetHandle.getSegmentName(), offset, sourceSegment);
         Timer timer = new Timer();
-        SortedSet<MultipartPartETag> partEtags = new TreeSet<>();
-        String targetPath = config.getPrefix() + targetHandle.getSegmentName();
-        String uploadId = client.initiateMultipartUpload(config.getBucket(), targetPath);
 
+        String targetPath = config.getPrefix() + targetHandle.getSegmentName();
         // check whether the target exists
         if (!doExists(targetHandle.getSegmentName())) {
             throw new StreamSegmentNotExistsException(targetHandle.getSegmentName());
         }
         // check whether the source is sealed
         SegmentProperties si = doGetStreamSegmentInfo(sourceSegment);
+        String sourcePath = config.getPrefix() + sourceSegment;
         Preconditions.checkState(si.isSealed(), "Cannot concat segment '%s' into '%s' because it is not sealed.",
                 sourceSegment, targetHandle.getSegmentName());
 
+        if (config.getSmallObjectSizeLimitForConcat() < si.getLength()) {
+            doConcatWithMultipartUpload(targetPath, sourceSegment, offset);
+            ExtendedS3Metrics.LARGE_CONCAT_COUNT.inc();
+        } else {
+            doConcatWithAppend(targetPath, sourcePath, offset, si.getLength());
+        }
+        // Now delete the source object.
+        client.deleteObject(config.getBucket(), sourcePath);
+
+        Duration elapsed = timer.getElapsed();
+        log.debug("Concat target={} source={} offset={} bytesWritten={} latency={}.", targetHandle.getSegmentName(), sourceSegment, offset, si.getLength(), elapsed.toMillis());
+
+        ExtendedS3Metrics.CONCAT_LATENCY.reportSuccessEvent(elapsed);
```

Comment:
_latency_ and _bytes_ are recorded only in case of success, in case of a failure in `deleteObject()` those will not be recorded. Don't they need to be recorded anyway?

Reply:
We never record these for failed requests.

Reply:
We can start reportFailEvent in case concat (or/and delete) failed. 
We may report success only after both concat and delete succeeded, otherwise report failure.

Reply:
Yes, we could do that. But this would be the only place in the code we currently do that. We can tackle this as a widespread change in another PR.

Reply:
Shall we delete the source segment through ExtendedS3Storage.delete(), instead of calling client.delete() directly? 
Going through that path, we have Tier2 DELETE metrics triggered.

Reply:
Yes, just overload the doDelete(SegmentHandle handle) with doDelete(String bucket, String prefix, String segmentName).

Reply:
For 0.7 let's update metrics for only successful requests. 
Let's track the issue as separate issue for 0.8